### PR TITLE
Fix constant error observations appearing in the logs

### PR DESCRIPTION
**Pull request by tofarr**
**Now checking that .gitignore exists. before trying to filter based on its contents.**

- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
I see the following constantly in the logs:
```
|21:13:47 - openhands:DEBUG: action_execution_server.py:166 - Running action:
    |FileReadAction(path='.gitignore', start=0, end=-1, thought='', action='read', security_risk=None)
    ...
    |21:13:47 - openhands:DEBUG: action_execution_server.py:168 - Action output:
    |**ErrorObservation**
    |File not found: /workspace/.gitignore. Your current working directory is /workspace.
    ...
```

The reason is that every time we call `/list-files` we try to filter based on the contents of `.gitignore` - so we get an error message every time we try to list files if there is no `.gitignore`.

This change checks that the `.gitignore` exists before trying to read it.

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:6e252fe-nikolaik   --name openhands-app-6e252fe   docker.all-hands.dev/all-hands-ai/openhands:6e252fe
```

### CODE REVIEW

**Comment by enyst** [comment]
For some reason, one of the image builds got stuck for over 4h. I restarted them.

File: openhands/server/routes/files.py [hunk 1]
```diff
@@ -77,10 +77,16 @@ async def list_files(request: Request, path: str | None = None):
 
     file_list = [f for f in file_list if f not in FILES_TO_IGNORE]
 
-    async def filter_for_gitignore(file_list, base_path):
-        gitignore_path = os.path.join(base_path, '.gitignore')
+    async def has_gitignore() -> bool:
+        file_list = await call_sync_from_async(runtime.list_files, '')
```
Comment (enyst) [inline openhands/server/routes/files.py:81]: I wonder if this is a bit heavy, to do twice, for every click. (first at line 68)

Could we avoid it somehow? Or it doesn't matter?

File: openhands/server/routes/files.py [hunk 2]
```diff
@@ -77,10 +77,16 @@ async def list_files(request: Request, path: str | None = None):
...
+        result = next((True for f in file_list if f == '.gitignore'), False)
+        return result
+
+    async def filter_for_gitignore(file_list):
+        if not await has_gitignore():
+            return file_list
         try:
-            read_action = FileReadAction(gitignore_path)
+            read_action = FileReadAction('.gitignore')
             observation = await call_sync_from_async(runtime.run_action, read_action)
```
Comment (enyst) [inline openhands/server/routes/files.py:90]: Just a thought, if there is no addition to the `stream` on this execution path (and it shouldn't be), could we just get/read the ErrorObs and treat it here directly?

Comment (tofarr) [reply openhands/server/routes/files.py:90]: With this approach, the logs would still contain:
```
|**ErrorObservation**
    |File not found: /workspace/.gitignore. Your current working directory is /workspace
```
The only ways around this that I can see are:

1. Live with this oddity.
2. Do not filter files using `.gitignore`
3. Check that the `.gitignore` exists. (The approach I took)
4. Update the `FileReadAction` with an optional flag to permit missing files and not yield an error observation.
5. Update the runtime to allow access to the FileStore and use this directly.

Comment (enyst) [reply openhands/server/routes/files.py:90]: Correct me if wrong, but isn't there also
6. Do not log stuff in the FileReadAction, but in its client code?

Or that sounds really bad? I don't think they are many... but I could be wrong.

File: openhands/server/routes/files.py (hunk 1 above)
Comment (rbren) [reply openhands/server/routes/files.py:81]: Yeah I agree we shouldn't do a `list_files` first--we should just try to access the file, and use error handling for FileNotFoundError

If that's causing logspam let's remove the logs upstream

File: openhands/server/routes/files.py (hunk 2 above)
Comment (enyst) [reply openhands/server/routes/files.py:90]: Actually, [this one](https://github.com/All-Hands-AI/OpenHands/blob/2f11634ccaeb3aa38893a994d53ff280bbf485d0/openhands/runtime/action_execution_server.py#L254) isn't doing it, it's the `run_action`.

Comment (rbren) [reply openhands/server/routes/files.py:90]: How about we use `runtime.read` directly instead of `run_action`

Comment (enyst) [reply openhands/server/routes/files.py:90]: Looks like it also does `run_action`. We could write one that doesn't, unless I'm missing something?

OK, but it's late and sorry, I no longer understand this problem in the first place: these are all in the sandbox, both the results of `list_files` and `.gitignore`, so why aren't we filtering during `runtime.list_files`?

**Review by enyst — CHANGES_REQUESTED** [review]
Thanks for doing this, the log spam is annoying and everyone will be better off when it's gone.

We have a good discussion in the comments, warts and confusions and all. I think hashing out the potential issues in comments makes a PR better, and I would love to see the final form here.

So I'll put a Request changes just to make sure... that it's not merged too fast. 😅 

(I noticed you merge PRs a little too fast quite often, before comments are addressed or even acknowledged. Sorry, but I think we can do this better, because sometimes that can break things and some of us really don't like to break things in `main` here.)

**Comment by tofarr** [comment]
I'm gonna punt on this one now as there does seem to be concerns about the additional overhead of a file existence check. Perhaps a more thorough refactor will allow a better solution

**Pull request closed without merging**